Feature/stock api permissions

chunky56 opened the pull request.

chunky56 committed:
VIMS-398 Add new stock card service call to support permissions

Add new call to get a product by stock card id to support permissions calls for getStockCardById API call.

chunky56 committed:
VIMS-398 Add permissions checks to stock card API calls

Add permissions checks to the three GET and one POST stock card API calls.

For the GET calls, before the call is made to the service, a list of rights is retrieved for the user, by facility and product. If the view stock on hand right is found in the list, then the service call is made, otherwise a 403 Forbidden error is returned. For the API call that gets multiple stock cards, if no stock cards exist, then a 404 Not Found is returned--otherwise, an empty stock card list is returned.

For the POST call, each stock event is checked, for a list of rights by user, facility and product. Each stock event needs to be checked, because different events might specify different products. The check returns a list of rights, and manage stock permission is searched in the list. If not found, then a 403 Forbidden error is returned. This means if any event has incorrect permissions, the whole API call does not go through.

Also, add test cases for the new code, and add Swagger API documentation for the changes.

chunky56 committed:
VIMS-398 Clean up some unnecessary statements

chunky56 committed:
VIMS-398 Fix test case for new permission code

chunky56 committed:
VIMS-398 Rename adjust and receive stock permissions to manage stock

chunky56 committed:
VIMS-398 Fix messages in controller to use message service

chunky56 committed:
VIMS-398 Fix test case to setup proper permissions

benleibert commented:
I started looking at this today, and will continue on Sunday or Monday.

benleibert commented on src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java:
```diff
@@ -398,5 +452,24 @@ private void removeEmptyLotsFromStockCard(StockCard stockCard)
         stockCard.setLotsOnHand(nonEmptyLots);
     }
 
+    private List<Right> getRightsForUserFacilityAndProductCode(Long userId, Long facilityId, String productCode)
```
It may be worth putting this code block (or even the entire getRightsForUserFacilityAndProductCode function) in a service in order to make it more reusable. ProgramService seems like perhaps the ideal spot. However, the code uses ProgramProductService as well as ProgramService, and the former already has a dependency on later. Putting this code in ProgramProductService would thus alleviate adding a circular-dependency to ProgramService.

benleibert commented on src/main/java/org/openlmis/stockmanagement/controller/StockCardController.java:
```diff
@@ -398,5 +452,24 @@ private void removeEmptyLotsFromStockCard(StockCard stockCard)
+    {
+        Facility facility = facilityRepository.getById(facilityId);
+
+        // Get programs by product code, through programProducts
```
It may be worth putting this code block (or even the entire getRightsForUserFacilityAndProductCode function) in a service in order to make it more reusable. ProgramService seems like perhaps the ideal spot. However, the code uses ProgramProductService as well as ProgramService, and the former already has a dependency on later. Putting this code in ProgramProductService would thus alleviate adding a circular-dependency to ProgramService.

chunky56 replied:
Since this deals with rights, it might be best to put it in the RoleRightsService then.

benleibert commented on src/main/java/org/openlmis/stockmanagement/service/StockCardService.java:
```diff
@@ -104,6 +105,10 @@ public StockCard getStockCardById(Long facilityId, Long stockCardId) {
     return repository.getStockCards(facilityId);
   }
 
+  public Product getProductByStockCardId(Long stockCardId) {
```
I've seen (and therefore repeated) this pattern in OpenLMIS, wherein a method added to a service does nothing other than wrap a call with the same method signature within a repository. The added code doesn't seem to add value, though, and thus doesn't seem ideal.

benleibert commented:
Everything looks good. There is a minor merge-conflict in StockCardController, though. If you’d like look at it, and consider my potential suggestions above, I’d be glad to merge the code.

chunky56 committed:
VIMS-398 Remove unnecessary service call and call repository directly

Service call simply is a wrapper for the repository call, so remove it and call repository directly.

chunky56 committed:
VIMS-398 Move get rights by user, facility and product to RoleRightsService

This helper function is generic, so it makes sense to move it to a service.

Also change calls to use the service call. Fix test case.

Commit:
Merge origin/develop into current branch.
